eoscompanion/routes.py

```diff
@@ -161,19 +160,419 @@ def desktop_id_to_app_id(desktop_id):
     '''Remove .desktop suffix from desktop_id.'''
     return os.path.splitext(desktop_id)[0]
 
+
+def yield_desktop_ids_from_feed_models(models):
+    '''For each model that has a desktop_id, yield the desktop_id.'''
+    for model in models:
+        try:
+            yield model.get_property('desktop-id')
+        except TypeError:
+            continue
+
+
+def app_infos_for_feed_models(models, callback):
+    '''Get a GDesktopAppInfo for each model source in models.
+
+    A source might have more than once model, so we deduplicate here.
+    '''
+    def _on_received_all_app_info_results(results):
+        '''Callback for when all results are received.
+
+        Check for errors and report them, ignoring sources that had errors.
+        Continue with the sources that do not have errors.
+        '''
+        app_infos = []
+
+        for result in results:
+            error, info = result
+
+            if error is not None:
+                logging.error('Encountered error in getting app info: %s', error)
+                continue
+
+            app_infos.append(info)
+
+        callback(None, app_infos)
+
+    def _load_application_info_thunk(desktop_id):
+        '''A thunk used here since we cannot use lambdas in for loops.'''
+        def _internal(callback):
```

Comment:
I'd call this with a different name, since `callback` is already the parameter for the top parent function